File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.py
def solution(numbers, target):
    dfs(numbers, 0, 0, target)
    
    return answer

def dfs(numbers, sum, i, target):
    global answer
    
    if i == len(numbers):
        if sum == target:
            answer += 1
        return
    
    dfs(numbers, sum + numbers[i], i + 1, target)
    dfs(numbers, sum - numbers[i], i + 1, target)
  
answer = 0

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.java
import java.util.*;

class Solution {
    public int solution(int[] numbers, int target) {
        Queue<Integer> leaves = new LinkedList<>();
        leaves.add(0);

        Queue<Integer> result = bfs(leaves, numbers);
        
        int answer = Collections.frequency(result, target);
        return answer;
    }
    
    public static Queue<Integer> bfs(Queue<Integer> leaves, int[] numbers) {
        for (int i = 0; i < numbers.length; i++) {
            Queue<Integer> temp = new LinkedList<>();
            while (!leaves.isEmpty()) {
                int leave = leaves.poll();
                temp.add(leave + numbers[i]);
                temp.add(leave - numbers[i]);
            }
            leaves = temp;
        }

        return leaves;
    }
}